### ISSUE TITLE
Support gradient checkpointing for functional models via the standard fit method

### ISSUE THREAD

**Issue by rivershah** [keras-team/keras#18116]
**System information**.

TensorFlow version (you are using): `2.12.0`
Are you willing to contribute it (Yes/No) : No

**Describe the feature and the current behavior/state**.

Describe the feature clearly here. Be sure to convey here why the requested feature is needed. Any brief description about the use-case would help.

The following feature excerpt is copied from `huggingface` documentation [here](https://huggingface.co/docs/transformers/v4.18.0/en/performance#gradient-checkpointing)

In order to compute the gradients during the backward pass all activations from the forward pass are normally saved. This can create a big memory overhead. Alternatively, one could forget all activations during the forward pass and recompute them on demand during the backward pass. This would however add a significant computational overhead and slow down training.

Gradient checkpointing strikes a compromise between the two approaches and saves strategically selected activations throughout the computational graph so only a fraction of the activations need to be re-computed for the gradients. See [this great article](https://medium.com/tensorflow/fitting-larger-networks-into-memory-583e3c758ff9) explaining the ideas behind gradient checkpointing.

**Will this change the current api? How?**

Yes, it will introduce a flag in `model.fit` such as `gradient_checkpointing=True`

**Who will benefit from this feature?**

Any user with large intermediate activations and tensors that won't fit in device memory. Sequence models can be very memory heavy and this allows scale up without introducing model parallelism complexity. 

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): No
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Comment by innat** [comment]
Grad checkpointing is a very attractive feature. It would be great to have official support for this.  Discussion: https://discuss.tensorflow.org/t/support-gradient-checkpointing-in-tensorflow-2/.

**Comment by innat** [comment]
@ianstenbit 
(not 100% sure), From TF, there is a method [recompute_grad](https://www.tensorflow.org/api_docs/python/tf/recompute_grad) for this. In EfficientDet official [repo](https://github.com/google/automl/tree/master/efficientdet#11-reducing-memory-usage-when-training-efficientdets-on-gpu), they used it for gradient-checkpiont, [code](https://github.com/google/automl/blob/f42454c547085e9855be6a0f73c4c9bae8cd71a2/efficientdet/tf2/efficientdet_keras.py#L427).

**Comment by ianstenbit** [comment]
Oh interesting -- thanks @innat, I wasn't aware of that. This may be feasible as a feature using `recompute_grad`. Let me sync with some folks that know a bit more.

**Comment by ianstenbit** [comment]
It sounds like this should be possible (probably through making a subclass of `Model` which uses a custom `train_step` and decorates `call` with this `recompute_grad` decorator.

This seems like a nice-to-have feature, but is not currently on our roadmap. I'll leave the issue open for now -- if any contributors are interested in experimenting with this approach, we could see how it goes.
